File: ADT/DSU/2316.cpp
class Solution {
public:
    vector<int> parent;
    vector<int> rank;

    int find(int x) {
        if (x == parent[x])
            return x;
        return parent[x] = find(parent[x]);
    }

    void Union(int x, int y) {
        int px = find(x);
        int py = find(y);
        if (px == py)
            return;
        if (rank[px] > rank[py]) {
            parent[py] = px;
        } else if (rank[px] < rank[py]) {
            parent[px] = py;
        } else {
            parent[px] = py;
            rank[py]++;
        }
    }

    long long countPairs(int n, vector<vector<int>>& edges) {
        parent.resize(n);
        rank.resize(n, 0);

        for (int i = 0; i < n; i++) {
            parent[i] = i;
        }

        for (auto& vec : edges) {
            int u = vec[0];
            int v = vec[1];
            Union(u, v);
        }

        unordered_map<int, int> mp;
        for(int i = 0; i < n;i++){
            int head = find(i);
            mp[head]++;
        }
        long long result = 0;
        long long remainingNodes = n;

        for (auto & it : mp) {
            long long size = it.second;
            result += (size) * (remainingNodes - size);
            remainingNodes -= size;
        }
        return result;
    }
};

File: ADT/DSU/2316.py
from collections import defaultdict


class Solution:
    def __init__(self):
        self.parent = []
        self.rank = []

    def find(self, x):
        if x == self.parent[x]:
            return x
        self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def Union(self, x, y):
        px = self.find(x)
        py = self.find(y)
        if px == py:
            return
        if self.rank[px] > self.rank[py]:
            self.parent[py] = px
        elif self.rank[px] < self.rank[py]:
            self.parent[px] = py
        else:
            self.parent[px] = py
            self.rank[py] += 1

    def countPairs(self, n: int, edges: List[List[int]]) -> int:
        self.parent = list(range(n))
        self.rank = [0] * n
        for u, v in edges:
            self.Union(u, v)
        mp = defaultdict(int)
        for i in range(n):
            head = self.find(i)
            mp[head] += 1

        result = 0
        remaining_nodes = n
        for size in mp.values():
            result += size * (remaining_nodes - size)
            remaining_nodes -= size
        return result
